score/apps/apps.go

```diff
@@ -133,3 +134,34 @@ func hasPodAntiAffinity(selfLables internal.MapLables, affinity *corev1.Affinity
 
 	return false
 }
+
+func statefulsetHasServiceName(allServices []ks.Service) func(statefulset appsv1.StatefulSet) (scorecard.TestScore, error) {
+	return func(statefulset appsv1.StatefulSet) (score scorecard.TestScore, err error) {
+		for _, service := range allServices {
+			if service.Service().Namespace != statefulset.Namespace ||
+				service.Service().Name != statefulset.Spec.ServiceName ||
+				service.Service().Spec.ClusterIP != "None" {
+				continue
+			}
+
+			labels := statefulset.Spec.Template.GetObjectMeta().GetLabels()
+			selector := service.Service().Spec.Selector
+
+			matchCount := 0
+			for selectorKey, selectorValue := range selector {
+				if labelValue, ok := labels[selectorKey]; ok && labelValue == selectorValue {
+					matchCount++
+				}
+			}
+
+			if matchCount == len(selector) {
+				score.Grade = scorecard.GradeAllOK
+				return
+			}
```

Comment:
Kubernetes has some built-in APIs for checking if a LabelSelector matches, which is mostly useful when dealing with LabelSelectorExpressions, that's not the case here, but I still think that this might be easier:

```suggestion
			// Construct a LabelSelector from the Service and check if it
			// matches the labels in the StatefulSet template
			labelSelector := &metav1.LabelSelector{
				MatchLabels: service.Service().Spec.Selector,
			}

			selector, labelErr := metav1.LabelSelectorAsSelector(labelSelector)
			if labelErr != nil {
				continue
			}

			lables := internal.MapLables(statefulset.Spec.Template.GetObjectMeta().GetLabels())

			if selector.Matches(internal.MapLables(lables)) {
				score.Grade = scorecard.GradeAllOK
				return
			}
```

Trying to implement a label matcher myself is the root cause behind #316 a few other bugs in the past.

Reply:
_Update:_ I noticed that this is a pattern that is a little bit annoying. I created a small helper in #319 for this use case.

If you rebase on the lastest master, you can use the following instead:

```go
			if internal.LabelSelectorMatchesLabels(
				service.Service().Spec.Selector,
				statefulset.Spec.Template.GetObjectMeta().GetLabels(),
			) {
				score.Grade = scorecard.GradeAllOK
				return
			}
```

Reply:
Thanks for your review and adding a cool helper! I've refactored this using the helper.
https://github.com/zegl/kube-score/pull/317/files#diff-fb1d2cb38fdd16f9dd4a24473c2aeac0R147-R153